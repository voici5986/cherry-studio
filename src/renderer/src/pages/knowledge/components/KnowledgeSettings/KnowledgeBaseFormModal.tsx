import { HStack } from '@renderer/components/Layout'
import { Menu, Modal, ModalProps } from 'antd'
import React, { useState } from 'react'
import styled from 'styled-components'

export interface PanelConfig {
  key: string
  label: string
  panel: React.ReactNode
}

interface KnowledgeBaseFormModalProps extends Omit<ModalProps, 'children'> {
  panels: PanelConfig[]
}

const KnowledgeBaseFormModal: React.FC<KnowledgeBaseFormModalProps> = ({ panels, ...rest }) => {
  const [selectedMenu, setSelectedMenu] = useState(panels[0]?.key)

  const menuItems = panels.map(({ key, label }) => ({ key, label }))
  const activePanel = panels.find((p) => p.key === selectedMenu)?.panel

  return (
    <StyledModal
      destroyOnHidden
      maskClosable={false}
      centered
      transitionName="animation-move-down"
      width="min(800px, 70vw)"
      styles={{
        body: { padding: 0, height: 550 },
        header: {
          padding: '10px 15px',
          borderBottom: '0.5px solid var(--color-border)',
          margin: 0,
          borderRadius: 0
        },
        content: {
          padding: 0,
          paddingBottom: 10,
          overflow: 'hidden'
        }
      }}
      {...rest}>
      <HStack height="100%">
        <LeftMenu>
          <StyledMenu
            defaultSelectedKeys={[selectedMenu]}
            mode="vertical"
            items={menuItems}
            onSelect={({ key }) => setSelectedMenu(key)}
          />
        </LeftMenu>
        <SettingsContentPanel>{activePanel}</SettingsContentPanel>
      </HStack>
    </StyledModal>
  )
}

const StyledModal = styled(Modal)`
  .ant-modal-title {
    font-size: 14px;
  }
  .ant-modal-close {
    top: 4px;
    right: 4px;
  }
`

const LeftMenu = styled.div`
  display: flex;
  height: 100%;
  border-right: 0.5px solid var(--color-border);
`

const SettingsContentPanel = styled.div`
  flex: 1;
  padding: 16px 16px;
  overflow-y: scroll;
`

const StyledMenu = styled(Menu)`
  width: 200px;
  padding: 5px;
  background: transparent;
  margin-top: 2px;
  border-inline-end: none !important;

  .ant-menu-item {
    height: 36px;
    color: var(--color-text-2);
    display: flex;
    align-items: center;
    border: 0.5px solid transparent;
    border-radius: 6px;
    margin-bottom: 7px;

    .ant-menu-title-content {
      line-height: 36px;
    }
  }
  .ant-menu-item-active {
    background-color: var(--color-background-soft) !important;
    transition: none;
  }
  .ant-menu-item-selected {
    background-color: var(--color-background-soft);
    border: 0.5px solid var(--color-border);
    .ant-menu-title-content {
      color: var(--color-text-1);
      font-weight: 500;
    }
  }
`

export default KnowledgeBaseFormModal
