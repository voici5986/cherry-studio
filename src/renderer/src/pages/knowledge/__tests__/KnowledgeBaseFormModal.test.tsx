import { fireEvent, render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import KnowledgeBaseFormModal, { PanelConfig } from '../components/KnowledgeSettings/KnowledgeBaseFormModal'

// Mock dependencies
const mocks = vi.hoisted(() => ({
  onCancel: vi.fn(),
  onOk: vi.fn()
}))

// Mock HStack component
vi.mock('@renderer/components/Layout', () => ({
  HStack: ({ children, ...props }: any) => (
    <div data-testid="hstack" {...props}>
      {children}
    </div>
  )
}))

// Mock antd components
vi.mock('antd', () => ({
  Modal: ({ children, open, title, onCancel, onOk, ...props }: any) =>
    open ? (
      <div data-testid="modal" data-title={title} {...props}>
        <div data-testid="modal-header">
          <span>{title}</span>
          <button type="button" data-testid="modal-close" onClick={onCancel}>
            ×
          </button>
        </div>
        <div data-testid="modal-body">{children}</div>
        <div data-testid="modal-footer">
          <button type="button" data-testid="modal-cancel" onClick={onCancel}>
            Cancel
          </button>
          <button type="button" data-testid="modal-ok" onClick={onOk}>
            OK
          </button>
        </div>
      </div>
    ) : null,
  Menu: ({ items, defaultSelectedKeys, onSelect, ...props }: any) => (
    <div data-testid="menu" data-default-selected={defaultSelectedKeys?.[0]} {...props}>
      {items?.map((item: any) => (
        <div
          key={item.key}
          data-testid={`menu-item-${item.key}`}
          onClick={() => onSelect?.({ key: item.key })}
          style={{ cursor: 'pointer' }}>
          {item.label}
        </div>
      ))}
    </div>
  )
}))

/**
 * 创建测试用的面板配置
 * @param overrides 可选的属性覆盖
 * @returns PanelConfig 数组
 */
function createPanelConfigs(overrides: Partial<PanelConfig>[] = []): PanelConfig[] {
  const defaultPanels: PanelConfig[] = [
    {
      key: 'general',
      label: 'General Settings',
      panel: <div data-testid="general-panel">General Settings Panel</div>
    },
    {
      key: 'advanced',
      label: 'Advanced Settings',
      panel: <div data-testid="advanced-panel">Advanced Settings Panel</div>
    }
  ]

  return defaultPanels.map((panel, index) => ({
    ...panel,
    ...overrides[index]
  }))
}

/**
 * 渲染 KnowledgeBaseFormModal 组件的辅助函数
 * @param props 可选的组件属性
 * @returns render 结果
 */
function renderModal(props: Partial<any> = {}) {
  const defaultProps = {
    open: true,
    title: 'Knowledge Base Settings',
    panels: createPanelConfigs(),
    onCancel: mocks.onCancel,
    onOk: mocks.onOk
  }

  return render(<KnowledgeBaseFormModal {...defaultProps} {...props} />)
}

describe('KnowledgeBaseFormModal', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  describe('basic rendering', () => {
    it('should match snapshot', () => {
      const { container } = renderModal()
      expect(container.firstChild).toMatchSnapshot()
    })

    it('should render modal when open is true', () => {
      renderModal({ open: true })

      expect(screen.getByTestId('modal')).toBeInTheDocument()
      expect(screen.getByTestId('hstack')).toBeInTheDocument()
      expect(screen.getByTestId('menu')).toBeInTheDocument()
    })

    it('should render first panel by default', () => {
      renderModal()

      expect(screen.getByTestId('general-panel')).toBeInTheDocument()
      expect(screen.queryByTestId('advanced-panel')).not.toBeInTheDocument()
    })

    it('should handle empty panels array', () => {
      renderModal({ panels: [] })

      expect(screen.getByTestId('modal')).toBeInTheDocument()
      expect(screen.getByTestId('menu')).toBeInTheDocument()
    })
  })

  describe('menu interaction', () => {
    it('should switch panels when menu item is clicked', () => {
      renderModal()

      // Initially shows general panel
      expect(screen.getByTestId('general-panel')).toBeInTheDocument()
      expect(screen.queryByTestId('advanced-panel')).not.toBeInTheDocument()

      // Click advanced menu item
      fireEvent.click(screen.getByTestId('menu-item-advanced'))

      // Should now show advanced panel
      expect(screen.queryByTestId('general-panel')).not.toBeInTheDocument()
      expect(screen.getByTestId('advanced-panel')).toBeInTheDocument()
    })

    it('should set default selected menu to first panel key', () => {
      const panels = createPanelConfigs()
      renderModal({ panels })

      const menu = screen.getByTestId('menu')
      expect(menu).toHaveAttribute('data-default-selected', panels[0].key)
    })

    it('should handle menu selection with custom panels', () => {
      const customPanels: PanelConfig[] = [
        {
          key: 'custom1',
          label: 'Custom Panel 1',
          panel: <div data-testid="custom1-panel">Custom Panel 1</div>
        },
        {
          key: 'custom2',
          label: 'Custom Panel 2',
          panel: <div data-testid="custom2-panel">Custom Panel 2</div>
        }
      ]

      renderModal({ panels: customPanels })

      // Initially shows first custom panel
      expect(screen.getByTestId('custom1-panel')).toBeInTheDocument()

      // Click second custom menu item
      fireEvent.click(screen.getByTestId('menu-item-custom2'))

      // Should now show second custom panel
      expect(screen.queryByTestId('custom1-panel')).not.toBeInTheDocument()
      expect(screen.getByTestId('custom2-panel')).toBeInTheDocument()
    })
  })

  describe('modal props', () => {
    const user = userEvent.setup()
    it('should pass through modal props correctly', () => {
      const customTitle = 'Custom Modal Title'
      renderModal({ title: customTitle })

      const modal = screen.getByTestId('modal')
      expect(modal).toHaveAttribute('data-title', customTitle)
    })

    it('should call onOk when ok button is clicked', async () => {
      renderModal()

      await user.click(screen.getByTestId('modal-ok'))
      expect(mocks.onOk).toHaveBeenCalledTimes(1)
    })
  })

  describe('edge cases', () => {
    it('should handle single panel', () => {
      const singlePanel: PanelConfig[] = [
        {
          key: 'only',
          label: 'Only Panel',
          panel: <div data-testid="only-panel">Only Panel</div>
        }
      ]

      renderModal({ panels: singlePanel })

      expect(screen.getByTestId('only-panel')).toBeInTheDocument()
      expect(screen.getByTestId('menu-item-only')).toBeInTheDocument()
    })

    it('should handle panel with undefined key gracefully', () => {
      const panelsWithUndefined = [
        {
          key: 'valid',
          label: 'Valid Panel',
          panel: <div data-testid="valid-panel">Valid Panel</div>
        }
      ]

      renderModal({ panels: panelsWithUndefined })

      expect(screen.getByTestId('valid-panel')).toBeInTheDocument()
    })
  })
})
