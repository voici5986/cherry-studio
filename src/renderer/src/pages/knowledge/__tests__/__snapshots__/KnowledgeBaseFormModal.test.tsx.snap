// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`KnowledgeBaseFormModal > basic rendering > should match snapshot 1`] = `
.c0 .ant-modal-title {
  font-size: 14px;
}

.c0 .ant-modal-close {
  top: 4px;
  right: 4px;
}

.c1 {
  display: flex;
  height: 100%;
  border-right: 0.5px solid var(--color-border);
}

.c3 {
  flex: 1;
  padding: 16px 16px;
  overflow-y: scroll;
}

.c2 {
  width: 200px;
  padding: 5px;
  background: transparent;
  margin-top: 2px;
  border-inline-end: none!important;
}

.c2 .ant-menu-item {
  height: 36px;
  color: var(--color-text-2);
  display: flex;
  align-items: center;
  border: 0.5px solid transparent;
  border-radius: 6px;
  margin-bottom: 7px;
}

.c2 .ant-menu-item .ant-menu-title-content {
  line-height: 36px;
}

.c2 .ant-menu-item-active {
  background-color: var(--color-background-soft)!important;
  transition: none;
}

.c2 .ant-menu-item-selected {
  background-color: var(--color-background-soft);
  border: 0.5px solid var(--color-border);
}

.c2 .ant-menu-item-selected .ant-menu-title-content {
  color: var(--color-text-1);
  font-weight: 500;
}

<div
  class="c0"
  data-testid="modal"
  data-title="Knowledge Base Settings"
  styles="[object Object]"
  transitionname="animation-move-down"
  width="min(800px, 70vw)"
>
  <div
    data-testid="modal-header"
  >
    <span>
      Knowledge Base Settings
    </span>
    <button
      data-testid="modal-close"
      type="button"
    >
      Ã—
    </button>
  </div>
  <div
    data-testid="modal-body"
  >
    <div
      data-testid="hstack"
      height="100%"
    >
      <div
        class="c1"
      >
        <div
          class="c2"
          data-default-selected="general"
          data-testid="menu"
          mode="vertical"
        >
          <div
            data-testid="menu-item-general"
            style="cursor: pointer;"
          >
            General Settings
          </div>
          <div
            data-testid="menu-item-advanced"
            style="cursor: pointer;"
          >
            Advanced Settings
          </div>
        </div>
      </div>
      <div
        class="c3"
      >
        <div
          data-testid="general-panel"
        >
          General Settings Panel
        </div>
      </div>
    </div>
  </div>
  <div
    data-testid="modal-footer"
  >
    <button
      data-testid="modal-cancel"
      type="button"
    >
      Cancel
    </button>
    <button
      data-testid="modal-ok"
      type="button"
    >
      OK
    </button>
  </div>
</div>
`;
